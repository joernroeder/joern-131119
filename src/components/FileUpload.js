import React, { useState, useRef, useCallback, useEffect } from 'react'
import PropTypes from 'prop-types'

import { useFilesDispatch, Actions } from '../store/FileStore'
import { useApiContext, ApiStatus } from '../api/ApiContext'

const FileUpload = (props) => {
  // get dispatch from store
  const dispatch = useFilesDispatch()

  const fileInput = useRef(undefined)

  // set up upload API
  const api = useApiContext()
  const { run: uploadFile, isLoading: isUploading } = api.uploadFile({
    onResolve: ({ data }) => {
      dispatch({
        type: Actions.ADD_FILE,
        payload: { file: data },
      })
    },
  })

  const { accept, maxFileSize, multiple, disabled } = props
  const { onValidationError } = props

  // TODO validators could also be passed via prop and some middleware pattern.
  const validateFile = (file) => {
    if (!accept.includes(file.type)) {
      throw new Error('The given file type is not allowed.')
    }

    if (file.size > maxFileSize) {
      throw new Error('The given file is to large.')
    }

    return true
  }

  const handleInputChange = (event) => {
    const { files } = event.target
    const [file] = files

    if (!file) {
      return
    }

    try {
      validateFile(file)
    } catch (error) {
      if (onValidationError) {
        onValidationError(error.message)
      }
      return
    }

    const data = new FormData()
    data.append('file', file)

    uploadFile(data)
  }

  const onOpenFileSelector = (event) => {
    event.preventDefault()

    if (disabled) {
      return
    }

    fileInput.current.click()
  }

  return (
    <form>
      <input
        type="file"
        ref={fileInput}
        accept={accept}
        multiple={multiple}
        onChange={handleInputChange}
        className="hidden"
        required
      />
      {React.cloneElement(props.children, {
        onClick: onOpenFileSelector,
        disabled: isUploading,
      })}
    </form>
  )
}

FileUpload.propTypes = {
  accept: PropTypes.arrayOf(PropTypes.string).isRequired,
  maxFileSize: PropTypes.number.isRequired,
  multiple: PropTypes.bool,
  disabled: PropTypes.bool,
  children: PropTypes.element.isRequired,
  onValidationError: PropTypes.func,
}

FileUpload.defaultProps = {
  multiple: false,
  disabled: false,
}

export default FileUpload
