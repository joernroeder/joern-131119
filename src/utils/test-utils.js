import React from 'react'

import '@testing-library/jest-dom/extend-expect'
import 'jest-prop-type-error'
import fetchMock from 'jest-fetch-mock'

import { render } from '@testing-library/react'

import { FilesProvider } from '../store/FileStore'
import { FilteredFilesProvider } from '../store/FilteredFileStore'
import { ApiProvider } from '../api/ApiContext'

global.fetch = fetchMock

const WithFilesProvider = ({ children }) => {
  return (
    <FilesProvider>
      <FilteredFilesProvider>{children}</FilteredFilesProvider>
    </FilesProvider>
  )
}

const WithApiAndFilesProvider = ({ children }) => {
  return (
    <ApiProvider>
      <WithFilesProvider>{children}</WithFilesProvider>
    </ApiProvider>
  )
}

const renderWithFilesProviders = (ui, options) => {
  return render(ui, { wrapper: WithFilesProvider, ...options })
}

const renderWithApiAndFileProviders = (ui, options) => {
  return render(ui, { wrapper: WithApiAndFilesProvider, ...options })
}

// re-export everything
export * from '@testing-library/react'

// override render method
export { renderWithFilesProviders, renderWithApiAndFileProviders }
