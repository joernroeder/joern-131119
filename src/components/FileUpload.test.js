import React from 'react'
import { renderWithApiAndFileProviders, fireEvent, act } from 'test-utils'

import axiosMock, { axiosInstance } from 'axios'
import FileUpload from './FileUpload'

const uploadProps = {
  accept: ['image/jpeg'],
  maxFileSize: 1e6, // 1mb
}

test('it should render correctly', () => {
  const { container } = renderWithApiAndFileProviders(
    <FileUpload {...uploadProps}>
      <button>Upload</button>
    </FileUpload>
  )

  expect(container).toMatchSnapshot()
})

test('the file input should be invisible', () => {
  const { container } = renderWithApiAndFileProviders(
    <FileUpload {...uploadProps}>
      <button>Upload</button>
    </FileUpload>
  )

  const input = container.querySelector('input')

  expect(input).toHaveClass('hidden')
})

test('a click on the child should open the file selector', () => {
  const { container, getByText } = renderWithApiAndFileProviders(
    <FileUpload {...uploadProps}>
      <button>Upload</button>
    </FileUpload>
  )

  const input = container.querySelector('input')
  act(() => {
    input.onclick = jest.fn()
    fireEvent.click(getByText('Upload'))
  })

  expect(input.onclick).toHaveBeenCalledTimes(1)
})

test('it should correctly call the onValidationHandler', () => {
  const onValidationError = jest.fn()

  const { container } = renderWithApiAndFileProviders(
    <FileUpload {...uploadProps} onValidationError={onValidationError}>
      <button>Upload</button>
    </FileUpload>
  )

  const input = container.querySelector('input')

  act(() => {
    const file = new File(['dummy content'], 'example.png', {
      type: 'image/png',
    })

    // @see https://github.com/testing-library/react-testing-library/issues/93#issuecomment-403887769
    Object.defineProperty(input, 'files', {
      value: [file],
    })

    fireEvent.change(input)
  })

  expect(onValidationError).toHaveBeenCalledTimes(1)
  expect(onValidationError.mock.calls[0][0]).toBe(
    'The given file type is not allowed.'
  )
})

test('it should correctly call the onValidationHandler if the file exceeds the limit', () => {
  const onValidationError = jest.fn()
  const maxFileSize = 9 // 9 bytes

  const { container } = renderWithApiAndFileProviders(
    <FileUpload
      {...uploadProps}
      maxFileSize={maxFileSize}
      onValidationError={onValidationError}
    >
      <button>Upload</button>
    </FileUpload>
  )

  const input = container.querySelector('input')

  act(() => {
    const file = new File(['0123456789'], 'ten-byte.jpg', {
      type: 'image/jpeg',
    })
    Object.defineProperty(input, 'files', {
      value: [file],
    })
    fireEvent.change(input)
  })

  expect(onValidationError).toHaveBeenCalledTimes(1)
  expect(onValidationError.mock.calls[0][0]).toBe('The given file is to large.')
})

/*
test('@current it should correctly call the axios instance', async () => {
  const onValidationError = jest.fn()
  const { container } = renderWithApiAndFileProviders(
    <FileUpload {...uploadProps} onValidationError={onValidationError}>
      <button>Upload</button>
    </FileUpload>
  )

  const input = container.querySelector('input')
  const file = new File(['dummy content'], 'example.png', {type: 'image/jpeg'})

  act(() => {
    // @see https://github.com/testing-library/react-testing-library/issues/93#issuecomment-403887769
    Object.defineProperty(input, 'files', {
      value: [file],
    })
  })

  act(() => {
    fireEvent.change(input)
  })

  console.log(onValidationError.mock.calls)

  console.log(axiosInstance.mock.calls)
})
 */
