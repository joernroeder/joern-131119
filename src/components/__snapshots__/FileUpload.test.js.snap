// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render correctly 1`] = `
<div>
  <form>
    <input
      accept="image/jpeg"
      class="hidden"
      required=""
      type="file"
    />
    <button>
      Upload
    </button>
  </form>
</div>
`;
